Angular build time optimization

kroeder opened the pull request:
Issue: https://github.com/storybooks/storybook/issues/3968

## What I did

**Note:** This is work-in-progress

- Renamed `loadTsConfig`  to `getTsLoaderOptions` because that is what it actually does (https://github.com/storybooks/storybook/commit/58010575192ec05cc061f7084f9cbb62240addb8#diff-aaa47dcc2ee48513f45210bb35a92d4cR1)
- Set `transpileOnly: true` as default option for `ts-loader`
- Added `fork-ts-checker-webpack-plugin` as a dependency in `app/angular`
- Use `ForkTsChecker` and set option `tsconfig` to the same that is used for `ts-loader` (using `getTsLoaderOptions`)

## Checklist

- [x] Set `transpileOnly: true` as default `ts-loader` option
- [x] Add `ForkTsChecker` to the list of plugins
- [x] Verify it works
- [x] Check, why `ForkTsChecker` does check `jest` (`ERROR in D:/04_Development/ng-projects/storybook/node_modules/@types/jest/index.d.ts(1103,9):` and many more in this file)
- [x] Update documentation
- [x] Update test-case `ts_config.test.js`

## How to test

Is this testable with Jest or Chromatic screenshots? **yes**
Does this need a new example in the kitchen sink apps? **no**
Does this need an update to the documentation? **yes**

If your answer is yes to any of these, please make sure to include it in your PR.

**Manual test:** 
Start `yarn storybook` in `examples/angular-cli/`

Using `ForkTsChecker` adds the folloing message when starting storybook
```
Starting type checking service...
Using 1 worker with 2048MB memory limit
```


For maintainers only: Please tag your pull request with at least one of the following:
`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

kroeder committed:
Integrated ForkTsCheckerWebpackPlugin

kroeder committed:
Fixed tsconfig path for ForkTsChecker

kroeder committed:
Updated test cases

kroeder commented:
Please note: I had `yarn test --core` errors right from the beginning when forking `master`. Is this a known thing?

ndelangen commented:
@igor-dv please verify if this is compatible with your preset PR

igor-dv commented:
I will review it soon (or when I will be bored of doing nothing in a vacation)

igor-dv commented on app/angular/src/server/ts_config.js:
```diff
@@ -13,13 +13,14 @@ function resolveTsConfig(tsConfigPath) {
-  if (!configFile) {
-    return {};
+  if (configFilePath) {
+    tsLoaderOptions.configFile = configFilePath;
```
don't we want to use this here too?

igor-dv commented on app/angular/src/server/wrapInitialConfig.js:
```diff
@@ -1,45 +1,55 @@
 import path from 'path';
```
please solve merge conflicts here, it moved to `framework-preset-angular`

igor-dv commented on app/angular/src/server/wrapInitialConfig.js:
```diff
@@ -1,45 +1,55 @@
+        path.resolve(__dirname, '..')
+      ),
+      new ForkTsCheckerWebpackPlugin({
+        tsconfig: tsLoaderOptions.configFile,
```
what will happen if there is not a `configFile` ?

igor-dv replied:
BTW, if `tslint` is always located in root, we can also add it here (just check if it exists).

kroeder commented:
Thanks for your comments! Do you think I should revert + change my approach to fit in into the new template feature?

igor-dv commented:
I don't think you need to revert. The preset has a similar structure, just move your changes there

igor-dv commented:
@kroeder, do you need help to finish this?

kroeder commented:
Normally I would say "no" but I spent hours trouble shooting why my project does not build, start, tests on master fail.. Even on a fresh checkout :( btw. I am the one asking lots of questions in slack. Im

kroeder committed:
Merge branch 'master' into angular-build-time-optimization

kroeder committed:
Merge branch 'master' into angular-build-time-optimization

# Conflicts:
#	app/angular/src/server/wrapInitialConfig.js

Created a function for ForkTsCheckerWebpackPlugin
Added tests for ForkTsCheckerWebpackPlugin

kroeder committed:
Updated instantiation behavior of ForkTsCheckerWebpackPlugin function; Updated tests

kroeder commented:
I created a seperate file for creating an instance of ForkTsChecker and added tests for it.

igor-dv commented on app/angular/src/server/framework-preset-fork-ts-checker-plugin.js:
```diff
@@ -0,0 +1,13 @@
+import ForkTsCheckerWebpackPlugin from 'fork-ts-checker-webpack-plugin';
```
Let's call this file `createForkTsCheckerPlugin` or something like this.

igor-dv commented on app/angular/src/server/framework-preset-fork-ts-checker-plugin.js:
```diff
@@ -0,0 +1,13 @@
+export default function(tsLoaderOptions) {
+  if (tsLoaderOptions && tsLoaderOptions.configFile) {
+    return new ForkTsCheckerWebpackPlugin({
+      tsconfig: tsLoaderOptions.configFile,
```
Don't you want to locate `tslint` as well? or it will be by default?

igor-dv commented on examples/angular-cli/package.json:
```diff
@@ -9,7 +9,7 @@
     "ng": "ng",
     "start": "ng serve",
-    "storybook": "npm run storybook:prebuild && start-storybook -p 9008 -s src/assets",
+    "storybook": "start-storybook -p 9008 -s src/assets",
```
I've fixed this, please merge from master and revert this

kroeder committed:
Merge branch 'master' into angular-build-time-optimization

kroeder committed:
reverted storybook angular-cli example start script

kroeder committed:
Removed @types/jasmine and @types/jasminewd2 to resolve a type collision with @types/jest

kroeder committed:
Renamed fork-ts-checker-plugin file

kroeder committed:
yarn.lock

kroeder committed:
Merge branch 'master' into angular-build-time-optimization

kroeder commented:
Checks are fine now, finally! I just need to integrate the cli stuff as well as the updated documentation for this PR. Anything else you want to add/change?

kroeder committed:
Added custom tsconfig.json due to the ForkTsChecker implementation

kroeder committed:
Added documentation

kroeder committed:
Updated documentation

kroeder commented:
I added `tsconfig.json` to the angular-cli `getstorybook` template and updated the documentation. 

How can I verify, that `getstorybook` works for angular-cli projects?

igor-dv reviewed (approved):
A few minor comments

We are almost there =)

igor-dv commented:
> How can I verify, that getstorybook works for angular-cli projects?

`getstorybook` is now called `storybook init` from the `@storybook/cli`

@Keraito, how are you testing CLI?

kroeder committed:
Merge branch 'master' into angular-build-time-optimization

kroeder commented on docs/src/pages/basics/guide-angular/index.md:
```diff
@@ -73,6 +73,27 @@ That'll load stories in `../src/stories/index.ts`.
 
+## Storybook TypeScript configuration
+
+**Note:** You need this only if you are using Storybook `>= 4.0.0-alpha.22`.
```
I can mention that but the whole guide is more or less a complete "manual installation" guide, isn't it?

https://storybook.js.org/basics/guide-angular/

`If you want to set up Storybook manually, this is the guide for you.`

Regarding `ForkTsCheckerWebpackPlugin` I'll add that.

kroeder committed:
Updated documentation

kroeder committed:
Updated documentation

kroeder committed:
yarn.lock

Keraito commented:
@igor-dv As said on discord, atm manually on dummy projects. Still trying to come up with something else that doesn't have the same issues as the removed CLI snapshots.

@kroeder what are you exactly trying to verify with the CLI?

kroeder committed:
Fixed grammar

Keraito commented:
> I want to check if a custom tsconfig.json is created on storybook init (only for angular cli projects)

The easiest for now is to probably check this manually for now. Tagging #1108 so I don't forget this.

> that a fresh project with storybook init still works out of the box.

You can use a smoke test for this. https://github.com/storybooks/storybook/tree/master/lib/cli/test/fixtures Add a example project here without Storybook installed. Running `yarn test --cli` will trigger the `storybook init` command, install deps and then start Storybook in it. It will give an error if the Storybook doesn't start successfully

kroeder commented:
I `npm link`ed `lib/cli` and created a fresh angular project. `storybook init` now creates a `tsconfig.json` inside `.storybook`. I also linked my changes from `@storybook/angular`. Everything seems to work but I get the following error 

```ERROR in ./src/styles.css (./node_modules/raw-loader!./node_modules/postcss-loader/lib??embedded!./node_modules/@storybook/core/node_modules/style-loader!./node_modules/css-loader??ref--17-1!./node_modules/@storybook/core/node_modules/postcss-loader/src??postcss!./src/styles.css)
Module build failed (from ./node_modules/postcss-loader/lib/index.js):
Syntax Error 

(2:1) Unknown word

  1 | 
> 2 | var content = require("!!../node_modules/css-loader/index.js??ref--17-1!../node_modules/@storybook/core/node_modules/postcss-loader/src/index.js??postcss!./styles.css");
    | ^
  3 | 
  4 | if(typeof content === 'string') content = [[module.id, content, '']];

 @ ./src/styles.css 2:14-325 21:1-42:3 22:19-330
 @ multi ./node_modules/@storybook/core/dist/server/config/polyfills.js ./node_modules/@storybook/core/dist/server/config/globals.js ./.storybook/config.js ./node_modules/webpack-hot-middleware/client.js?reload=true ./src/styles.css
i ｢wdm｣: Failed to compile.
``` 

I don't know if it's related to my PR - anyone knows something about this error?

igor-dv commented:
I think it's not related 👍
